### DISCUSSION TITLE
use NIRI_SOCKET variable in systemd user services

### DISCUSSION THREAD

**Discussion by TheAngusMcFire** [YaLTeR/niri#527] [Q&A]
Hi everybody, 

I try to start 2 services which need to communicate with niri as systemd user services.
One is the xwayland-satellite and one is a rust service which uses the niri-ipc crate to communicate with niri.

Both systemd files look like that.
```
[Unit]
Description=Script Daemon for xwayland-satellite 
After=graphical-session.target
Requisite=graphical-session.target
PartOf=graphical-session.target

[Service]
Type=simple
ExecStart=/usr/bin/xwayland-satellite :1000
Restart=on-failure
```

The problem I am having is that the services start to fast, so that the NIRI_SOCKET variable is not yet set.
niri-ipc reports the missing env var.
If I restart the 2 services everything is working fine, so I think systemd is not waiting long enough  before it starts those 2 services.

Has anybody done something similar or does know how I can configure the systemd service to wait for the env var to be present?

System: Arch Linux
DisplayManager: GDM

**Comment by YaLTeR** [comment]
Are you on NixOS? It seems to have some problem with graphical-session.target dependencies, so you could try replacing them with niri.service dependencies.

**Reply by TheAngusMcFire** [reply]
I am on arch linux

**Reply by YaLTeR** [reply]
How are you running niri?

**Reply by TheAngusMcFire** [reply]
I tried replacing it with Requires=niri.service and it produced the same behaviour

**Reply by TheAngusMcFire** [reply]
I start niri via the gdm display manager

**Reply by YaLTeR** [reply]
Hm, should work then... Try replacing all three lines, just put niri.service instead of graphical-session.target

**Reply by TheAngusMcFire** [reply]
Oh, thank you very much this actually works now:

```
[Unit]
Description=Script Daemon for xwayland-satellite 
PartOf=niri.service
After=niri.service
Requisite=niri.service

[Service]
Type=simple
ExecStart=/usr/bin/xwayland-satellite :1000
Restart=on-failure
```

**Reply by TheAngusMcFire** [reply]
Hm ok no, sorry it is very flaky It seems it only works if the timing is right.
It only worked 2 of 5 times.

**Reply by TheAngusMcFire** [reply]
They are as symlinks in the niri.service.wants folder, thats how they get started

**Reply by YaLTeR** [reply]
Yeah, that's normal. Not sure, I haven't had any failures with the same setup on Fedora.

**Reply by TheAngusMcFire** [reply]
how does systemd know that niri is started? Is it guaranteed that the environment was published before the service goes into active state?

**Reply by YaLTeR** [reply]
It uses Type=notify, which means that it waits for a message from niri, which niri sends after publishing the environment.

**Reply by TheAngusMcFire** [reply]
I see, thank you for the help, I think I have to debug a bit for this one :see_no_evil:

**Comment by TheAngusMcFire** [comment]
I kept having this problem, but I just had time for some additional debugging, and it seems that the NIRI_SOCKET variable contains the wrong or old NIRI_SOCKET value.

NIRI_SOCKET of the service:  /run/user/1000/niri.wayland-1.2668771.sock
NIRI_SOCKET of the current instance: /run/user/1000/niri.wayland-1.2685705.sock

Is this a systemd bug maybe?

**Reply by TheAngusMcFire** [reply]
It seams that the services are not stopped when niri is stopped, what makes it more strange: if I switch from gdm to another tty, the service is suddenly stopped...

**Comment by YaLTeR** [comment]
FYI, this is the expected order of events (copying from my message in https://github.com/sodiboo/niri-flake/issues/348):

1. `niri-session` runs and starts `niri.service`
2. `niri.service` ExecStart-s `/usr/bin/niri --session`, then waits for a notification (`Type=notify`)
3. niri imports WAYLAND_DISPLAY and whatnot into the systemd environment:
    * https://github.com/YaLTeR/niri/blob/7b6fa1285475896ad8dbc2f4b11099b0c036e4b7/src/main.rs#L209
    * https://github.com/YaLTeR/niri/blob/7b6fa1285475896ad8dbc2f4b11099b0c036e4b7/src/main.rs#L271-L305
4. niri notifies systemd that it is ready: https://github.com/YaLTeR/niri/blob/7b6fa1285475896ad8dbc2f4b11099b0c036e4b7/src/main.rs#L224
5. systemd now proceeds to start graphical-session.target (since niri.service says `Before=graphical-session.target`). At this point the environment is filled in and everything is ready
6. systemd starts everything that has `After=graphical-session.target`, `Requisite=graphical-session.target`, i.e. example swaybg.service from the wiki page

**Comment by TheAngusMcFire** [comment]
Thanks for the detailed explanation. The problem was not related to niri at all.
The services did not get stopped when niri got stopped for some reason. I added a systemctl --user restart <service> as `spawn-at-startup` this worked perfectly.
Thanks for the help.
